src/simulators.jl

```diff
@@ -438,3 +438,121 @@ function B_step!(s, dt_eff, acceleration_vector, compute_forces::Bool, n_threads
 end
+
+"""
+    TemperatureREMD(; <keyword arguments>)
+
+A simulator for a parallel temperature replica exchange (TREX) simulation on a [`ReplicaSystem`](@ref). More information on this algorithm can be found in [Sugita Y., Okamoto Y. 1999](https://doi.org/10.1016/S0009-2614(99)01123-9).
+The corresponding [`ReplicaSystem`](@ref) should have the same number of replicas as the number of temperatures in the simulator.
+
+arguments:
+- `dt::S`: the time step of the simulation.
+- `temperatures::TP`: the temperatures corresponding to the replicas.
+- `simulators::ST`: individual simulators for simulating each replica.
+- `exchange_time::ET`: the time interval between replica exchange attempt.
+"""
+struct TemperatureREMD{N, T, S, DT, TP, ST, ET}
+    dt::DT
+    temperatures::TP
+    simulators::ST
+    exchange_time::ET
+end
+
+function TemperatureREMD(;
+                dt,
+                temperatures,
+                simulators,
+                exchange_time,
+                kwargs...)
+    S = eltype(simulators)
+    T = eltype(temperatures)
+    N = length(temperatures)
+    DT = typeof(dt)
+    TP = typeof(temperatures)
+    ET = typeof(exchange_time)
+    if length(simulators) != length(temperatures)
+        throw(ArgumentError("Number of temperatures must match number of simulators"))
+    end
+    if exchange_time <= dt
+        throw(ArgumentError("Exchange time must be greater than the time step"))
+    end
+    simulators = Tuple(simulators[i] for i in 1:N)
+    ST = typeof(simulators)
+    
+    return TemperatureREMD{N, T, S, DT, TP, ST, ET}(dt, temperatures, simulators, exchange_time)
+end
+
+function simulate!(sys::ReplicaSystem,
+                    sim::TemperatureREMD,
+                    n_steps::Int;
+                    assign_velocities::Bool=false,
+                    rng=Random.GLOBAL_RNG,
+                    n_threads::Int=Threads.nthreads())
+    if sys.n_replicas != length(sim.simulators)
+        throw(ArgumentError("Number of replicas in ReplicaSystem and simulators in TemperatureREMD do not match."))
+    end
+
+    if n_threads > sys.n_replicas
+        thread_div = equal_parts(n_threads, sys.n_replicas)
+    else # pass 1 thread per replica
+        thread_div = equal_parts(sys.n_replicas, sys.n_replicas)
+    end
+
+    # calculate n_cycles and n_steps_per_cycle from dt and exchange_time
+    n_cycles = convert(Int64, (n_steps * sim.dt) ÷ sim.exchange_time)
+    cycle_length = n_steps ÷ n_cycles
+    remaining_steps = n_steps % n_cycles
+
+    if assign_velocities
+        for i in eachindex(sys.replicas)
+            random_velocities!(sys.replicas[i], sim.temperatures[i])
+        end
+    end
+
+    for cycle=1:n_cycles
+        @sync for idx in eachindex(sim.simulators)
+            Threads.@spawn Molly.simulate!(sys.replicas[idx], sim.simulators[idx], cycle_length; n_threads=thread_div[idx])
+        end
+
+        if cycle != n_cycles
+            n = rand(1:sys.n_replicas)
```

Comment:
I think all pairs should be considered each cycle with another loop otherwise the exchange rate is too low. Following the Gromacs policy (https://manual.gromacs.org/current/reference-manual/algorithms/replica-exchange.html) of exchanging odd pairs on odd cycles and even pairs on even cycles sounds like a good idea.

Reply:
I have changed the exchange scheme to the one above, please check if it is fine. After the change, the number of exchanges have increased from about 5-6 to 12-16 for the above simulation.

Reply:
Looks good.